base/iterators.jl

```diff
@@ -1476,4 +1476,97 @@ only(x::NamedTuple) = throw(
     ArgumentError("NamedTuple contains $(length(x)) elements, must contain exactly 1 element")
 )
 
+"""
+    unfold(f, initialstate, [eltype])
+
+Iterable object that generates values from an initial state and a transition
+function `f(state)`. The function must follow the same rules as `iterate`.
+It returns either `(newvalue, newstate)` or `nothing`, in which case the
+sequence ends.
+
+The optional parameter `eltype` can specify the type of the generated values.
+
+See also: [`iterate`](@ref)
+
+!!! compat "Julia 1.9"
+    This function was added in Julia 1.9.
+
+# Examples
+
+```jldoctest
+julia> fib = Iterators.unfold((1,1)) do (a,b)
+           a, (b, a+b)
+       end;
+
+julia> reduce(hcat, Iterators.take(fib, 7))
+1×7 Matrix{Int64}:
+ 1  1  2  3  5  8  13
+
+julia> frac(c, z=0.0im) = Iterators.unfold((c, z), ComplexF64) do (c, z)
+           if real(z * z') < 4
+               z, (c, z^2 + c)
+           else
+               nothing
+           end
+       end;
+
+julia> [count(Returns(true), frac(-0.835-0.2321im, (k+j*im)/6)) for j in -4:4, k in -8:8]
+9×17 Matrix{Int64}:
+  2   2   2   3   3   3   5  41   8   4   3   3   2   2   2   2   1
+  2   3   5   4   5   8  20  11  17  23   4   3   3   3   2   2   2
+  4  10  17  12   7  56  18  58  33  22   6   5   4   5   4   3   2
+ 26  56  15  13  18  23  13  14  27  46   8   9  16  12   8   4   3
+ 10   7  62  17  16  23  11  12  39  12  11  23  16  17  62   7  10
+  3   4   8  12  16   9   8  46  27  14  13  23  18  13  15  56  26
+  2   3   4   5   4   5   6  22  33  58  18  56   7  12  17  10   4
+  2   2   2   3   3   3   4  23  17  11  20   8   5   4   5   3   2
+  1   2   2   2   2   3   3   4   8  41   5   3   3   3   2   2   2
+```
+
+# Extended help
+
+The interface for `f` is very similar to the interface required by `iterate`, but `unfold` is simpler to use because it does not require you to define a type. You can use this to your advantage when prototyping or writing one-off iterators.
+
+You may want to define an iterator type instead if you'd like to define `IteratorSize` or `IteratorEltype`; for readability; or if you want to dispatch on the type of your iterator.
```

Comment:
```suggestion
You may want to define an iterator type instead if you'd like to define `IteratorSize`; or for readability; or if you want to dispatch on the type of your iterator.
```

This version of `unfold` lets you specify the eltype.

Reply:
The suggestion here is because `unfold` allows specifying the Eltype, so defining your own IteratorEltype method isn't likely to be a reason to define your own iterator type instead of using `unfold`.

Not that important, just thought this was clearer.

I don't have permission to merge this PR. I think it's mostly ready, tho I do think it would be a significant improvement to use a keyword arg as mentioned previously. I asked a couple of months ago for this to be reviewed at triage, which I think is still how stuff gets merged.